### DISCUSSION TITLE
How to execute a plan from a `duckdb::LogicalOperator`

### DISCUSSION THREAD

**Discussion by drin** [duckdb/duckdb#11098] [Q&A] [answered]
Hello! I am trying to figure out how to execute a query plan given a duckdb::LogicalOperator.

Based on the [internals overview][doc-internals], I feel like `LogicalOperator` -> `PhysicalOperator` (via [PhysicalPlanGenerator::CreatePlan][repo-plangen]) is the "preferred path" to generating a physical plan that can be executed by either:
1. A [duckdb::Executor][repo-exec], or
2. A [duckdb::PipelineExecutor][repo-pipexec]

However, a `duckdb::Relation` can be executed very easily by calling `Relation::Execute()` (which calls [ClientContext::Execute(Relation)][repo-execrel], then `ClientContext::PendingQueryInternal`, then `ClientContext::ExecutePendingQueryInternal`).

Is there a similarly straightforward approach to execute a `PhysicalOperator` without having to do the full setup of an `Executor` or `PipelineExecutor`?

Thank you for any help!

[doc-internals]: https://duckdb.org/internals/overview

[repo-exec]: https://github.com/duckdb/duckdb/blob/v0.10.0/src/include/duckdb/execution/executor.hpp#L33
[repo-pipexec]: https://github.com/duckdb/duckdb/blob/v0.10.0/src/include/duckdb/parallel/pipeline_executor.hpp#L36

[repo-plangen]: https://github.com/duckdb/duckdb/blob/v0.10.0/src/include/duckdb/execution/physical_plan_generator.hpp#L40-L42
[repo-execinit]: https://github.com/duckdb/duckdb/blob/v0.10.0/src/include/duckdb/execution/executor.hpp#L47-L48
[repo-execrel]: https://github.com/duckdb/duckdb/blob/v0.10.0/src/main/client_context.cpp#L1094-L1103

**Comment by drin** [comment]
I figured it out using duckdb::Executor with a relatively small amount of code, though significantly more than a single call (such as `ClientContext::Execute(Relation&)`).

I will do a bit of documentation and add here when it's ready.

**Reply by drin** [reply, marked as answer]
The short answer is:
1. `LogicalOperator -> PhysicalOperator` via PhysicalPlanGenerator
2. Create a `PreparedStatementData` (needed for next step)
3. Create a `PhysicalResultCollector`
4. Create an `Executor` and initialize it with the result collector (which `isa` `PhysicalOperator`) via `Executor::Initialize`
5. Use the `Executor` to execute tasks; I used `Executor::ExecuteTask` directly after `Executor::WorkOnTasks` gave me a segfault (and executing tasks directly is how I figured out that a result collector is necessary).

**Comment by Bouncner** [comment]
Hey @drin, I would like to do something similar. Is there a chance that the code you are working with is somewhere public?

**Reply by Tishj** [reply]
This is all extrapolated from the duckdb source code.

A good starting point to look at would be `ClientContext::CreatePreparedStatementInternal` as this is where the PreparedStatementData is created, it uses the Planner to turn the SQLStatement into a LogicalPlan, which you can replace with your manually crafted LogicalPlan

**Reply by drin** [reply]
Tishj is right, I did extrapolate from the duckdb source code, though it took me awhile to figure it all out. To give more details on the "short answer" I originally posted:

0. To get a `LogicalOperator` from a `Relation` I constructed a `SelectStatement` from the `Relation` instance then ran it through a `Planner`: [TranspilePlanMessage][fn-transpile]. I still kept the `Relation` around to make it easy to construct a `PreparedStatementData`.

1. A `duckdb::PhysicalOperator` can be obtained by passing a `LogicalOperator` to a `PhysicalPlanGenerator`: [TranslateLogicalPlan][fn-translate].

2. To create a [PreparedStatementData][cls-stmtdata], you can reference [BindingFnTranslateMohair][logic-plandata]:
```cpp
duckdb::Relation& plan_rel = ...
auto plan_data = make_shared_ptr<PreparedStatementData>(StatementType::SELECT_STATEMENT);

// For us to further build PreparedStatementData (probably affects our ResultCollector)
for (auto &column : plan_rel.Columns()) {
  plan_data->types.emplace_back(column.Type());
  plan_data->names.emplace_back(column.Name());
}

plan_data->plan = <duckdb::PhysicalOperator>
```

And here are where helper classes are defined:
* [FnDataSubstraitTranslation][cls-fndata]
* [3 Aliases][alias-plans] for duckdb "plans": `DuckSystemPlan`(`duckdb::Relation`), `DuckLogicalPlan`(`duckdb::LogicalOperator`), `DuckPhysicalPlan`(`duckdb::PhysicalOperator`). These are for me to abstract over whatever classes other engines may use.

3. Once you construct a `PreparedStatementData`, you can use it to create a `PhysicalResultCollector`: ([GetResultCollector][call-getcollector])

4. Create and use a `duckdb::Executor` ([DuckDBExecutor::Execute][fn-execute]) for execution:
* [ShouldKeepExecuting][fn-shouldexecute] shows how to check the result (`duckdb::PendingExecutionTask`) of executing a task (`duckdb::PipelineTask`)
* My wrapper class is simple, but it is defined here: [DuckDBExecutor][cls-duckexecutor]

The above code is probably still not straightforward since I plan to target a variety of query engines (e.g. datafusion, tiledb), but hopefully it helps a bit. Feel free to ask more questions, or I can try and write a simplified recipe you can reference. Also, if there are better ways to get `LogicalOperator` and `PhysicalOperator`, etc. please share! My code was so far just to prototype the pieces to see if I could even access the various code paths.



[fn-transpile]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/engine_duckdb.cpp#L94-L109
[fn-translate]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/engine_duckdb.cpp#L125-L129

[fn-execute]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/engine_duckdb.cpp#L160-L184
[fn-shouldexecute]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/engine_duckdb.cpp#L134-L158
[cls-duckexecutor]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/include/engine_duckdb.hpp#L208-L216
[cls-fndata]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/include/engine_duckdb.hpp#L107-L117
[alias-plans]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/include/engine_duckdb.hpp#L74-L76

[call-getcollector]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/engine_duckdb.cpp#L165

[logic-plandata]: https://github.com/drin/duckdb-substrait/blob/dabc3dbd169253e673f3274638e4e88cf659a3af/src/substrait_extension.cpp#L319-L327
[cls-stmtdata]: https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/main/prepared_statement_data.hpp#L25-L39
